KSPCommunityFixes/Performance/MemoryLeaks.cs

```diff
@@ -580,6 +694,155 @@ static bool ModuleScienceExperiment_OnAwake_Prefix()
+                }
+
+                __instance.protoPartSnapshot.partRef = null;
+            }
+
+            // clear some additional containers to limit the impact of leaks and make reports less noisy
+            __instance.modules?.modules?.Clear();
+            __instance.children?.Clear();
+            __instance.parent = __instance.potentialParent = null;
+            __instance.internalModel = null;
+
+#if DEBUG
+            if (FlightGlobals.objectToPartUpwardsCache.Values.Contains(__instance))
+            {
+                Debug.LogError($"Destroying part {__instance.GetInstanceID()} while it's still in the objectToPartUpwardsCache");
+            }
+#endif
+        }
+
+        static void Part_ReleaseAutoStruts_Postfix(Part __instance)
+        {
+            __instance.autoStrutCacheVessel = null;
+            __instance.autoStrutVessel = null;
+        }
+
+        // PartSets are awful as noted in OnSceneUnloaded.  But vessel unloading is also a good time to clean these up instead of waiting for scene switch
+
+        // general unhooking of stuff when unloading a vessel.  Note that vessel gameobjects are not usually destroyed until the vessel itself is
+        static void Vessel_Unload_Postfix(Vessel __instance)
+        {
+            __instance.rootPart = null;
+            __instance.referenceTransformPart = null;
+            __instance.referenceTransformPartRecall = null;
+            __instance.GroundContacts?.Clear();
+            __instance._suspensionLoadBalancer?.suspensionModules?.Clear();
+            __instance.dockingPorts?.Clear();
+
+            // the flight integrator holds onto a lot of references, which will be repopulated when the vessel is loaded again
+            // clear them all now
+
+            if (__instance._fi != null)
+            {
+                __instance._fi.partRef = null;
+                __instance._fi.partThermalDataList?.Clear();
+                __instance._fi.partThermalDataListSkin?.Clear();
+                __instance._fi.partThermalDataCount = 0;
+                __instance._fi.partCount = 0;
+                __instance._fi.compoundParts?.Clear();
+                __instance._fi.occlusionConv?.Clear();
+                __instance._fi.occlusionSun?.Clear();
+                __instance._fi.occlusionBody?.Clear();
+                __instance._fi.overheatModules?.Clear();
+                __instance._fi.previewModules?.Clear();
+                __instance._fi.occludersConvection = null;
+                __instance._fi.occludersSun = null;
+                __instance._fi.occludersBody = null;
+                __instance._fi.occludersBodyCount = 0;
+                __instance._fi.occludersConvectionCount = 0;
+                __instance._fi.occludersSunCount = 0;
+            }
+
+            // not sure about
+            // objectUnderVessel
+
+            FlightGlobals.ResetObjectPartUpwardsCache();
+            FlightGlobals.ResetObjectPartPointerUpwardsCache();
```

Comment:
Not ideal to reset that cache here since that will often cause it to be immediately rebuilt in every case a debris or another vessel gets out of loading range. Ideally we could be a bit smarter by waiting for next frame and only removing destroyed parts from those caches, but probably not worth the trouble, and I agree this is a potentially quite bad leak.